### PR TITLE
perf(univariate): optimize eval and pow backends

### PR DESCRIPTION
**Pull request by eliasjudin**
## Summary

Optimize the univariate evaluation and power backends by:
- replacing the public raw and canonical evaluation backends with Horner-style evaluation while keeping the existing `eval`/`eval₂` entrypoints;
- replacing the public raw power backend with exponentiation by squaring;
- re-establishing the lawful raw power theorems and the downstream `toPoly`, quotient, and bivariate bridge theorems needed by the canonical API.

This PR adds proofs autoformalised by @Aristotle-Harmonic.

## Motivation

The univariate layer previously evaluated raw polynomials through explicit sums of powers and powered raw polynomials through repeated multiplication. Those definitions were semantically correct but asymptotically wasteful, and Phase 2 called for moving both surfaces to the standard Horner/squaring backends without changing the public canonical behavior.

## Mathematical content

For evaluation, the new backend proves that the Horner fold agrees with the previous sum-of-powers semantics and transports that equality through the existing `toPoly` bridge, canonical `CPolynomial` API, and the bivariate `toPoly` bridge.

For powers, the public raw backend is replaced by exponentiation by squaring. The raw proof layer is updated at the mathematically correct lawful boundary, with lawful raw power theorems identifying the fast backend with repeated multiplication and preserving the canonical semiring behavior on `CPolynomial` and `QuotientCPolynomial`.

The PR also adds univariate regression tests covering zero, constant, monomial, trailing-zero, and nontrivial evaluation cases, together with canonical power regressions through the existing univariate test surface.

## Testing

- `./scripts/lint-style.sh CompPoly/Univariate/Raw/Ops.lean CompPoly/Univariate/Raw/Proofs.lean CompPoly/Univariate/Basic.lean CompPoly/Univariate/Quotient/Core.lean CompPoly/Univariate/ToPoly/Core.lean CompPoly/Univariate/ToPoly/Impl.lean CompPoly/Bivariate/ToPoly.lean tests/CompPolyTests/Univariate/Basic.lean tests/CompPolyTests/Univariate/Eval.lean tests/CompPolyTests.lean`
- `lake build`
- `lake test`
- `rg -n "sorry|admit|axiom|unsafe" CompPoly.lean CompPoly/**/*.lean tests/**/*.lean`

Refs #186

Co-authored-by: Aristotle (Harmonic) <aristotle-harmonic@harmonic.fun>

### CODE REVIEW

**Review by dhsorens — COMMENTED** [review]
hey @eliasjudin , thank you for this PR! this is definitely something that we want, sorry for the delay as I've been traveling. We've recently merged #199 that adds Horner's method. this PR makes the swap - can you merge in the main branch into this?

to be clear - from main we want `eval₂ => eval₂Naive` and `eval₂Horner => eval₂` as you have it here

File: CompPoly/Univariate/Raw/Proofs.lean [hunk 1]
```diff
@@ -30,12 +30,98 @@ variable (p q r : CPolynomial.Raw R)
...
+    induction p using Array.recOn
+    simp_all +decide
+    induction ‹List R› using List.reverseRecOn <;>
+      simp_all +decide [Finset.sum_range_succ]
+    simp_all +decide [Finset.sum_range, List.zipIdx_append]
+
+/-
+`eval₂` equals the list sum of mapped `zipIdx` pairs.
+-/
+theorem eval₂_eq_list_sum (f : R →+* S) (x : S) (p : CPolynomial.Raw R) :
```
Comment (dhsorens) [inline CompPoly/Univariate/Raw/Proofs.lean:107]: I think this theorem is covered in `eval₂_eq_eval₂_naive` (and currently on main as `eval₂_eq_eval₂Horner`) and unused anywhere else - may not need it?
